### PR TITLE
[Darwin] Adding stub API to store data

### PR DESCRIPTION
**Pull request by woody-apple**
Adding stub data for this client metadata. Needs to be hooked up to storage later.

### CODE REVIEW

File: src/darwin/Framework/CHIP/MTRDevice.h [hunk 1]
```diff
@@ -240,6 +240,49 @@ typedef NS_ENUM(NSUInteger, MTRDeviceState) {
                                       completion:(MTRDeviceOpenCommissioningWindowHandler)completion
     MTR_AVAILABLE(ios(17.0), macos(14.0), watchos(10.0), tvos(17.0));
 
+/**
+ *
+ * These set of functions allow clients to set metadata on a MTRDevice level, as well as per endpoint as well.
```
[the remark this answers is not in the document]
Comment (woody-apple) [reply src/darwin/Framework/CHIP/MTRDevice.h]: I'll note this, when an endpoint is deleted, we should drop it.

File: src/darwin/Framework/CHIP/MTRDevice.h [hunk 2]
```diff
@@ -240,6 +240,49 @@ typedef NS_ENUM(NSUInteger, MTRDeviceState) {
...
+- (id<NSSecureCoding>)clientDataForKey:(NSString *)key MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * Set client metadata for a key. The value must conform to NSSecureCoding
+ *
+ * @param key           NSString * for the key to store the value as
+ * @param value         id <NSSecureCoding> for the value to store
+ */
+- (void)setClientDataForKey:(NSString *)key value:(id<NSSecureCoding>)value MTR_NEWLY_AVAILABLE;
```
[the remark this answers is not in the document]
Comment (woody-apple) [reply src/darwin/Framework/CHIP/MTRDevice.h:265]: I was debating it, but opted to keeping it hidden until needed. Open to suggestions.

File: src/darwin/Framework/CHIP/MTRDevice.h [hunk 3]
```diff
@@ -240,6 +240,90 @@ typedef NS_ENUM(NSUInteger, MTRDeviceState) {
...
+ *   • Supported client data object types are currently only:
+ *         NSData, NSString, NSArray, NSDictionary, NSNumber
+ */
+
+/**
+ *
+ * List of all client data types supported
+ *
+ */
+- (NSArray *)supportedClientDataClasses MTR_NEWLY_AVAILABLE;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.h:258]: This should either document what the things in the NSArray are, or have it be `NSArray<Whatever>`, no?

File: src/darwin/Framework/CHIP/MTRDevice.h [hunk 4]
```diff
@@ -240,6 +240,90 @@ typedef NS_ENUM(NSUInteger, MTRDeviceState) {
...
+
+/**
+ *
+ * List of all client data keys stored
+ *
+ */
+- (NSArray * _Nullable)clientDataKeys MTR_NEWLY_AVAILABLE;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.h:265]: Likewise, this should either document what the things in the NSArray are, or have it be `NSArray<Whatever>`, no?

Also, when does this return nil?

File: src/darwin/Framework/CHIP/MTRDevice.h [hunk 5]
```diff
@@ -240,6 +240,90 @@ typedef NS_ENUM(NSUInteger, MTRDeviceState) {
...
+ * @param key           NSString * for the key to store the value as
+ */
+- (void)removeClientDataForKey:(NSString *)key MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * List of all client data keys stored
+ *
+ */
+- (NSArray * _Nullable)clientDataKeysForEndpointID:(NSNumber *)endpointID MTR_NEWLY_AVAILABLE;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.h:297]: See above.

File: src/darwin/Framework/CHIP/MTRDevice.h [hunk 6]
```diff
@@ -240,6 +240,90 @@ typedef NS_ENUM(NSUInteger, MTRDeviceState) {
...
+ * Retrieve client metadata for a key, returns nil if no value is set
+ *
+ * @param key           NSString * for the key to store the value as
+ * @param endpointID    NSNumber * for the endpoint to associate the metadata with
+ */
+- (id<NSSecureCoding> _Nullable)clientDataForKey:(NSString *)key endpointID:(NSNumber *)endpointID MTR_NEWLY_AVAILABLE;
+
+/**
+ *
+ * Set client metadata for a key. The value must conform to NSSecureCoding.
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDevice.h:310]: https://github.com/project-chip/connectedhomeip/pull/30730#discussion_r1410850766 does not seem to have been addresse?

**Pull request merged**